test/distributed/tensor/parallel/test_tp_examples.py

```diff
@@ -152,6 +158,147 @@ def test_mlp_inference(self):
         with torch.inference_mode():
             self._test_mlp_inference(device_mesh)
 
+    @with_comms
+    @skip_unless_torch_gpu
+    @parametrize("is_seq_parallel", [True, False])
+    def test_transformer_training(self, is_seq_parallel=False):
+        # Step 1: Initialize single-gpu models and optimizers.
+
+        model_args = ModelArgs(
+            # Disable dropout in the test since we cannot reproduce the same random
+            # behaviors when comparing single-gpu models with multi-gpu models.
+            dropout_p=0.0,
+            # TODO: Weight_tying works fine under inputs and models of small size.
+            # Test error would surpass the allowed limit when tuning up model size.
+            # Need to investigate if this is normal, e.g. due to precision issues.
+            weight_tying=False,
+        )
+
+        # Reset random seeds to ensure two models have the same initialization.
+        torch.manual_seed(5)
+        model = Transformer(model_args).to(self.device_type)
+        model_tp = deepcopy(model)
+        self._check_module(model, model_tp)
+
+        # Step 2: Set up and execute the parallelize plan to shard the test model
+        # onto the device mesh.
+
+        parallelize_plan = {}
+        device_mesh = DeviceMesh(self.device_type, torch.arange(0, NUM_DEVICES))
+
+        # Parallelize the embedding submodules.
+        parallelize_plan["tok_embeddings"] = ColwiseParallel(
+            output_layouts=Shard(1),
+        ) if is_seq_parallel else ColwiseParallel(output_layouts=Replicate())
+        parallelize_plan["pos_embeddings"] = ColwiseParallel(
+            output_layouts=Shard(0),
+        ) if is_seq_parallel else ColwiseParallel(output_layouts=Replicate())
+
+        # Parallelize the attention and feed forward submodules.
+        for i in range(model_args.n_layers):
+            if is_seq_parallel:
+                # parallelize_plan[f"layers.{i}.attention"] = PrepareModuleInput(
+                #     input_layouts=Shard(1),
+                #     desired_input_layouts=Replicate(),
+                # )
+                parallelize_module(
+                    model_tp.layers[i].attention,
+                    device_mesh,
+                    PrepareModuleInput(input_layouts=Shard(1), desired_input_layouts=Replicate())
+                )
+            parallelize_module(model_tp.layers[i].attention.wq, device_mesh, ColwiseParallel())
```

Comment:
hmmm. I think we should have a separate parallelize_plan for each transformer layer defined as a dict of fqn -> parallel styles. Calling parallelize_module so many times is not preferred I think. 

Each layer we can have a parallelize_plan then a single `parallelize_module` call, and then the first two embedding layer and the final projection layer can call `paralllelize_module`.

Reply:
Oh I see. I misunderstood your previous comment; in fact I also thought this way of calling is not good. I wonder in general what principles user should follow when parallelizing their model. E.g. should one always try to group submodules within a bigger module into a single parallelize_plan, as long as the plan does grow very big? What are the concerns of not having a super giant single plan, like what I did earlier?

Reply:
I think a super giant plan also works, it's just that when we have multiple layers with the same plan, things become a bit redundant, so I was thinking for transformer that have repeated sharding plan, per-layer sharding plan makes things more clean